File: app/src/main/java/com/jann_luellmann/thekenapp/view/ListTableViewAdapter.java
package com.jann_luellmann.thekenapp.view;

import android.content.Context;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.LinearLayout;
import android.widget.TextView;

import com.evrencoskun.tableview.adapter.AbstractTableAdapter;
import com.evrencoskun.tableview.adapter.recyclerview.holder.AbstractViewHolder;
import com.jann_luellmann.thekenapp.R;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;

public class ListTableViewAdapter extends AbstractTableAdapter<ColumnHeader, RowHeader, Cell> {

    @NonNull
    @Override
    public AbstractViewHolder onCreateCellViewHolder(@NonNull ViewGroup parent, int viewType) {
        View layout = LayoutInflater.from(parent.getContext())
                .inflate(R.layout.table_view_cell_layout, parent, false);

        return new CellViewHolder(layout);
    }

    @Override
    public void onBindCellViewHolder(@NonNull AbstractViewHolder holder, @Nullable Cell cellItemModel, int columnPosition, int rowPosition) {
        CellViewHolder viewHolder = (CellViewHolder) holder;
        viewHolder.cellTextview.setText(cellItemModel.getData());

        // Resize
        viewHolder.cellContainer.getLayoutParams().width = LinearLayout.LayoutParams.WRAP_CONTENT;
        viewHolder.cellTextview.requestLayout();
    }

    @NonNull
    @Override
    public AbstractViewHolder onCreateColumnHeaderViewHolder(@NonNull ViewGroup parent, int viewType) {
        View layout = LayoutInflater.from(parent.getContext())
                .inflate(R.layout.table_view_column_header_layout, parent, false);

        return new ColumnHeaderViewHolder(layout);
    }

    @Override
    public void onBindColumnHeaderViewHolder(@NonNull AbstractViewHolder holder, @Nullable ColumnHeader columnHeaderItemModel, int columnPosition) {
        ColumnHeaderViewHolder columnHeaderViewHolder = (ColumnHeaderViewHolder) holder;
        columnHeaderViewHolder.columnHeaderTextView.setText(columnHeaderItemModel.getData());

        // Resize
        columnHeaderViewHolder.columnHeaderContainer.getLayoutParams().width = LinearLayout.LayoutParams.WRAP_CONTENT;
        columnHeaderViewHolder.columnHeaderTextView.requestLayout();
    }

    @NonNull
    @Override
    public AbstractViewHolder onCreateRowHeaderViewHolder(@NonNull ViewGroup parent, int viewType) {
        View layout = LayoutInflater.from(parent.getContext())
                .inflate(R.layout.table_view_row_header_layout, parent, false);

        return new RowHeaderViewHolder(layout);
    }

    @Override
    public void onBindRowHeaderViewHolder(@NonNull AbstractViewHolder holder, @Nullable RowHeader rowHeaderItemModel, int rowPosition) {
        RowHeaderViewHolder rowHeaderViewHolder = (RowHeaderViewHolder) holder;
        rowHeaderViewHolder.rowHeaderTextView.setText(rowHeaderItemModel.getData());
    }

    @NonNull
    @Override
    public View onCreateCornerView(@NonNull ViewGroup parent) {
        return LayoutInflater.from(parent.getContext())
                .inflate(R.layout.table_view_corner_layout, parent, false);
    }

    @Override
    public int getColumnHeaderItemViewType(int position) {
        return 0;
    }

    @Override
    public int getRowHeaderItemViewType(int position) {
        return 0;
    }

    @Override
    public int getCellItemViewType(int position) {
        return 0;
    }

    class CellViewHolder extends AbstractViewHolder {

        final LinearLayout cellContainer;
        final TextView cellTextview;

        public CellViewHolder(View itemView) {
            super(itemView);
            cellContainer = itemView.findViewById(R.id.cell_container);
            cellTextview = itemView.findViewById(R.id.cell_data);

        }
    }

    class ColumnHeaderViewHolder extends AbstractViewHolder {

        final LinearLayout columnHeaderContainer;
        final TextView columnHeaderTextView;

        public ColumnHeaderViewHolder(View itemView) {
            super(itemView);
            columnHeaderContainer = itemView.findViewById(R.id.column_header_container);
            columnHeaderTextView = itemView.findViewById(R.id.column_header_textView);
        }
    }

    class RowHeaderViewHolder extends AbstractViewHolder {

        final TextView rowHeaderTextView;

        public RowHeaderViewHolder(View itemView) {
            super(itemView);
            rowHeaderTextView = itemView.findViewById(R.id.row_header_textView);
        }
    }
}

File: app/src/main/java/com/jann_luellmann/thekenapp/view/ListTableViewAdapter.kt
package com.jann_luellmann.thekenapp.view

import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.TextView
import androidx.constraintlayout.widget.ConstraintLayout
import com.evrencoskun.tableview.TableView
import com.evrencoskun.tableview.adapter.AbstractTableAdapter
import com.evrencoskun.tableview.adapter.recyclerview.holder.AbstractViewHolder
import com.jann_luellmann.thekenapp.R
import com.jann_luellmann.thekenapp.util.Util

class ListTableViewAdapter : AbstractTableAdapter<ColumnHeader?, RowHeader?, Cell?>() {

    var tableView: TableView? = null
    private var rowHeadWidth = 0

    private val resizableViewHolders: MutableList<BaseViewHolder> = mutableListOf()

    override fun onCreateCellViewHolder(parent: ViewGroup, viewType: Int): AbstractViewHolder {
        val layout: View = LayoutInflater.from(parent.context)
            .inflate(R.layout.table_view_cell_layout, parent, false)
        return CellViewHolder(layout)
    }

    override fun onBindCellViewHolder(
        holder: AbstractViewHolder,
        cellItemModel: Cell?,
        columnPosition: Int,
        rowPosition: Int
    ) {
        val viewHolder = holder as CellViewHolder
        viewHolder.cellTextview.text = cellItemModel!!.data.toString()
    }

    override fun onCreateColumnHeaderViewHolder(
        parent: ViewGroup,
        viewType: Int
    ): AbstractViewHolder {
        val layout: View = LayoutInflater.from(parent.context)
            .inflate(R.layout.table_view_column_header_layout, parent, false)
        val viewHolder = ColumnHeaderViewHolder(layout)
        resizableViewHolders.add(viewHolder)
        return viewHolder
    }

    override fun onBindColumnHeaderViewHolder(
        holder: AbstractViewHolder,
        columnHeaderItemModel: ColumnHeader?,
        columnPosition: Int
    ) {
        val columnHeaderViewHolder = holder as ColumnHeaderViewHolder
        columnHeaderViewHolder.columnHeaderTextView.text = columnHeaderItemModel!!.data.toString()
    }

    override fun onCreateRowHeaderViewHolder(parent: ViewGroup, viewType: Int): AbstractViewHolder {
        return when (viewType) {
            ROW_HEADER_TEXT -> {
                val view: View = LayoutInflater.from(parent.context)
                    .inflate(R.layout.table_view_row_header_layout, parent, false)
                RowHeaderViewHolder(view)
            }
            else -> {
                val view: View = LayoutInflater.from(parent.context)
                    .inflate(R.layout.table_view_customer_row_header_layout, parent, false)
                RowHeaderCustomerViewHolder(view)
            }
        }
    }

    override fun onBindRowHeaderViewHolder(
        holder: AbstractViewHolder,
        rowHeaderItemModel: RowHeader?,
        rowPosition: Int
    ) {
        rowHeaderItemModel?.let {
            if (holder is RowHeaderViewHolder) {
                holder.bind(rowHeaderItemModel)
                rowHeadWidth = holder.cellContainer.layoutParams.width
            } else if (holder is RowHeaderCustomerViewHolder) {
                holder.bind(it)
                rowHeadWidth = holder.cellContainer.layoutParams.width
            }
        }
    }

    override fun onCreateCornerView(parent: ViewGroup): View {
        return LayoutInflater.from(parent.context)
            .inflate(R.layout.table_view_corner_layout, parent, false)
    }

    override fun getColumnHeaderItemViewType(position: Int): Int {
        return 0
    }

    override fun getRowHeaderItemViewType(position: Int): Int {
        return when (mRowHeaderItems[position]) {
            is RowHeaderCustomer -> ROW_HEADER_CUSTOMER
            else -> ROW_HEADER_TEXT
        }
    }

    override fun getCellItemViewType(position: Int): Int {
        return 0
    }

    internal open inner class BaseViewHolder(itemView: View) : AbstractViewHolder(itemView) {
        val cellContainer: ConstraintLayout = itemView.findViewById(R.id.cell_container)
    }

    internal inner class CellViewHolder(itemView: View) : BaseViewHolder(itemView) {
        val cellTextview: TextView = itemView.findViewById(R.id.cell_data)
    }

    internal inner class ColumnHeaderViewHolder(itemView: View) : BaseViewHolder(itemView) {
        val columnHeaderTextView: TextView = itemView.findViewById(R.id.column_header_textView)
    }

    internal inner class RowHeaderViewHolder(itemView: View) : BaseViewHolder(itemView) {
        private val rowHeaderTextView: TextView = itemView.findViewById(R.id.rowHeaderTextView)

        fun bind(rowHeaderItemModel: RowHeader) {
            rowHeaderTextView.text = rowHeaderItemModel.data.toString()
        }
    }

    internal inner class RowHeaderCustomerViewHolder(itemView: View) : BaseViewHolder(itemView) {
        private val rowHeaderButton: TextView = itemView.findViewById(R.id.rowHeaderButton)

        fun bind(rowHeaderItemModel: RowHeader) {
            rowHeaderButton.text = rowHeaderItemModel.data.toString()
            rowHeaderButton.setOnClickListener {
                rowHeaderItemModel.onClick()
            }
        }
    }

    companion object {
        val ROW_HEADER_CUSTOMER = 0
        val ROW_HEADER_TEXT = 1
    }
}